### ISSUE TITLE
No results when performing a walk starting on a leaf OID

### ISSUE THREAD

**Issue by danielnelson** [soniah/gosnmp#170]
When performing a bulk walk on a leaf node no results are returned, this is different than net-snmp and can be confusing.  Performing a get directly returns the value correctly
```
$ snmpbulkwalk -v2c -c public 127.0.0.1:161 SNMPv2-MIB::sysName.0
SNMPv2-MIB::sysName.0 = STRING: loaner
```

Here is a example program that does:
- get .1.3.6.1.2.1.1.5.0
- bulkwalk .1.3.6.1.2.1.1.5.0
- bulkwalk2 .1.3.6.1.2.1.1.5

```
func main() {
	client := &gosnmp.GoSNMP{
		Target:         "localhost",
		Port:           uint16(161),
		Community:      "public",
		MaxRepetitions: 10,
		Version:        gosnmp.Version2c,
		Timeout:        time.Duration(10) * time.Second,
	}

	err := client.Connect()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Connect() err: %v", err)
		return
	}
	defer client.Conn.Close()

	// Using Get
	result, err := client.Get([]string{".1.3.6.1.2.1.1.5.0"})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Get() err: %v", err)
		return
	}

	octets, err := json.MarshalIndent(result, "get: ", "\t")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Marshal() err: %v", err)
		return
	}
	fmt.Println("get:", string(octets))

	// Using BulkWalk
	err = client.BulkWalk(".1.3.6.1.2.1.1.5.0", func(dataUnit gosnmp.SnmpPDU) error {
		octets, err := json.MarshalIndent(dataUnit, "bulkwalk: ", "\t")
		if err != nil {
			fmt.Fprintf(os.Stderr, "Marshal() err: %v", err)
			return nil
		}

		fmt.Println("bulkwalk:", string(octets))
		return nil
	})

	if err != nil {
		fmt.Fprintf(os.Stderr, "BulkWalk() err: %v", err)
		return
	}

	// Using BulkWalk on parent
	err = client.BulkWalk(".1.3.6.1.2.1.1.5", func(dataUnit gosnmp.SnmpPDU) error {
		octets, err := json.MarshalIndent(dataUnit, "bulkwalk2: ", "\t")
		if err != nil {
			fmt.Fprintf(os.Stderr, "Marshal() err: %v", err)
			return nil
		}

		fmt.Println("bulkwalk2:", string(octets))
		return nil
	})

	if err != nil {
		fmt.Fprintf(os.Stderr, "BulkWalk() err: %v", err)
		return
	}
}
```

Output: 
```
get: {
get:    "Version": 1,
get:    "Variables": [
get:            {
get:                    "Name": ".1.3.6.1.2.1.1.5.0",
get:                    "Type": 4,
get:                    "Value": "bG9hbmVy",
get:                    "Logger": {}
get:            }
get:    ],
get: }
bulkwalk2: {
bulkwalk2:      "Name": ".1.3.6.1.2.1.1.5.0",
bulkwalk2:      "Type": 4,
bulkwalk2:      "Value": "bG9hbmVy",
bulkwalk2:      "Logger": {}
bulkwalk2: }
```

The bulkwalk directly on the leaf node returns no values, even though the value is returned from the SNMP agent.  Here is a tcpdump of the transaction:
```
18:19:47.516564 IP localhost.52113 > localhost.snmp:  GetRequest(28)  system.sysName.0
18:19:47.516747 IP localhost.snmp > localhost.52113:  GetResponse(34)  system.sysName.0="loaner"
18:19:47.516998 IP localhost.52113 > localhost.snmp:  GetBulk(28)  N=0 M=10 system.sysName.0
18:19:47.517196 IP localhost.snmp > localhost.52113:  GetResponse(255)  system.sysLocation.0="Sitting on the Dock of the Bay" system.sysServices.0=72 system.8.0=0 system.9.1.2.1=S:snmpMPDMIB.snmpMPDMIBConformance.snmpMPDMIBCompliances.1 system.9.1.2.2=S:snmpUsmMIB.usmMIBConformance.usmMIBCompliances.1 system.9.1.2.3=S:snmpFrameworkMIB.snmpFrameworkMIBConformance.1.1 system.9.1.2.4=S:1 system.9.1.2.5=S:16.2.2.1 system.9.1.2.6=49 system.9.1.2.7=ip
18:19:47.517252 IP localhost.52113 > localhost.snmp:  GetRequest(28)  system.sysName.0
18:19:47.517334 IP localhost.snmp > localhost.52113:  GetResponse(34)  system.sysName.0="loaner"
18:19:47.517379 IP localhost.52113 > localhost.snmp:  GetBulk(27)  N=0 M=10 system.sysName
18:19:47.517450 IP localhost.snmp > localhost.52113:  GetResponse(253)  system.sysName.0="loaner" system.sysLocation.0="Sitting on the Dock of the Bay" system.sysServices.0=72 system.8.0=0 system.9.1.2.1=S:snmpMPDMIB.snmpMPDMIBConformance.snmpMPDMIBCompliances.1 system.9.1.2.2=S:snmpUsmMIB.usmMIBConformance.usmMIBCompliances.1 system.9.1.2.3=S:snmpFrameworkMIB.snmpFrameworkMIBConformance.1.1 system.9.1.2.4=S:1 system.9.1.2.5=S:16.2.2.1 system.9.1.2.6=49
```

The walk is terminated in this conditional https://github.com/soniah/gosnmp/blob/master/walk.go#L64 because the v.Name and rootOid are equal.

I believe this is has been discussed in #78, and it seems like 37a797ddd084e3ab998cebd6748df2f1054643df by @wdreeveii was meant to, among other things, address this issue.

**Comment by soniah** [comment]
Thanks Daniel for reporting this. Would you mind providing packet dumps? That way I can write some tests to handle this correctly. Several dumps of `snmpbulkwalk` calling against different targets - things that have a few OIDs but not too many; as well as the expected terminal output.

See https://github.com/soniah/gosnmp#packet-captures for how to do this.

**Comment by soniah** [comment]
ACK! Thanks for doing the uploads, most people don't.

**Comment by soniah** [comment]
Hmmm... time to write some code....

**Comment by soniah** [comment]
Fixed with bd2e184d912ba28f9374fd46b9032a350a55aedb

Thanks @thebinary!
